Data consent updates and privacy improvements (Fixes #14213)

alexgibson opened the pull request:
## One-line summary

> [!NOTE]
> I apologize for this PR being large, with many things to test. Whoever reviews I owe you a beverage of your choice at all-hands 🍺 🍸 🍹 

This PR implements the following changes:

- Adds explicit opt-in for data collection in EU/EAA countries only on specific URLs.
- Omits non-essential cookies and analytics on URLs in EU/EAA countries where opt-in is not requested.
- Adds global data preferences page and link in main footer.
- Implements Global Privacy Control (GPC) as an additional opt-out signal to Do Not Track (DNT).

## Significant changes and points to review

- I tried to break down testing instructions per page type / analytics function in order to make things easier. Sorry for the list being long, but it hopefully covers everything.
- I also tried to separate out consent logic from analytics/attribution logic, to try and make both tests and code easier to follow.

## Issue / Bugzilla link

#14213

## Testing

Successful test run: https://github.com/mozilla/bedrock/actions/runs/9113965630

> [!IMPORTANT]
> Make sure to clear cookies before / after each test.

---

### Cookie settings page (EU):

URL: https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/?geo=de

- [x] When `moz-consent-pref` consent cookie is **NOT** set, default form values should equal **"I Do Not Agree"**.
- [x] When clicking "Save Changes", `moz-consent-pref` consent cookie **IS** set.
  - [x] Updating form values and clicking "Save Changes" should update the existing cookie.
  - [x] Refreshing the page should show the values stored in the cookie reflected in the form.
  

---

### Cookie settings page (rest-of-world):

URL: https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/

- [x] When `moz-consent-pref` consent cookie is **NOT** set, default form values should equal **"I Agree"**.

---

### Firefox Campaign Pages (EU)

URL: https://www-demo1.allizom.org/de/firefox/challenge-the-default/?geo=de

#### With GPC/DNT enabled:

- [x] Consent banner is **NOT** shown. 
- [x] GTM script is **NOT** loaded.
  - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.

#### With GCP/DNT disabled:

- [x] Consent banner **IS** shown.
- [x] Clicking "Accept" closes banner.
  - [x] GTM script **IS** loaded.
    - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
  - [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.
  - [x] Refreshing page persists consent choice and analytics behavior.
  - Open: https://www-demo1.allizom.org/de/firefox/download/thanks/?geo=de
    - [x] GTM script **IS** loaded.
    - [x] `attribution_code` and `attribution_sig` parameters **ARE** added to the Firefox auto-download URL.
- [x] Clicking "Reject" closes banner.
  - [x] GTM script **NOT** loaded.
  - [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.
  - [x] Refreshing page persists consent choice and behavior.

#### GTM test:

- [x] Test page load / download events fire as expected in GTM when accepting cookies.

---

### Firefox Campaign Pages (rest-of-world)

URL: https://www-demo1.allizom.org/de/firefox/challenge-the-default/

#### With GPC/DNT enabled:

- [x] GTM script is **NOT** loaded.
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.

#### With GPC/DNT disabled:

- [x] GTM script **IS** loaded.
  - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.
 
With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] GTM script is **NOT** loaded after page refresh.
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set (you may need to clear these if they were already set previously before opt-out).

With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] GTM script **IS** loaded after page refresh.
  - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.

#### GTM test:

- [x] Test page load / download events fire as expected in GTM.

---

### Firefox product pages (EU):

URL: https://www-demo1.allizom.org/en-US/firefox/new/?geo=de

- [x] Consent banner is **NOT** shown.
- [x] GTM (Google Tag Manager) script is **NOT** loaded.
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.

---

### Firefox product pages pages (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/firefox/new/

#### With GPC/DNT enabled:

- [x] GTM script is **NOT** loaded.
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.

#### With GCP/DNT disabled:

- [x] GTM script **IS** loaded.
  - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.
- Click download
  - [x] `attribution_code` and `attribution_sig` parameters **ARE** added to the Firefox auto-download URL.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] GTM script is **NOT** loaded.
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set (you may need to clear these if they were already set previously before opt-out).

With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] GTM script **IS** loaded.
  - [x] Verify `window.dataLayer` contains `core-datalayer-loaded` and `page-id-loaded` events. 
- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.
- Click download
  - [x] `attribution_code` and `attribution_sig` parameters **ARE** added to the Firefox auto-download URL.

---

### Firefox RTAMO flow (same logic applies for EU and rest-of-world)

URL: https://www-demo1.allizom.org/en-US/firefox/download/thanks/?s=direct&utm_source=test&utm_campaign=test

#### With GPC/DNT enabled:

- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies are **NOT** set.
- [x] Firefox auto-download triggers **WITHOUT** `attribution_code` and `attribution_sig` parameters.

#### With GPC/DNT disabled:

- [x] `moz-stub-attribution-code` and `moz-stub-attribution-sig` cookies **ARE** set.
- [x] Firefox auto-download triggers **WITH** `attribution_code` and `attribution_sig` parameters.

---

### FxA form (EU)

URL: https://www-demo1.allizom.org/en-US/firefox/?geo=de&utm_source=test&utm_campaign=test

Note: since these are form fields, you might need to do a hard refresh between each test to reset the form.

- [x] Consent banner is **NOT** shown (since the URL is not in the allow-list).
- [x] `utm_source=test` and `utm_campaign=test` are **NOT** passed through to the form `fxa-email-form-utm-source` and `fxa-email-form-utm-campaign` hidden fields.
- [x] FxA flow params (`flow_id`, `device_id`) are **NOT** added to hidden fields.
- [x] When filling out the form, email **IS* still passed through to FxA as expected.

When using Firefox on desktop:

- [x] `<input type="hidden" name="context" value="fx_desktop_v3">` **IS** added to the form.

---

### FxA form (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/firefox/?utm_source=test&utm_campaign=test

Note: since these are form fields, you might need to do a hard refresh between each test to reset the form.

#### By default: (does not require DNT/GPC)

- [x] `utm_source=test` and `utm_campaign=test` **ARE** passed through to the form `fxa-email-form-utm-source` and `fxa-email-form-utm-campaign` hidden fields.
- [x] FxA flow params (`flow_id`, `device_id`) **ARE** added to hidden fields.
- [x] When filling out the form, email **IS* still passed through to FxA as expected.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] `utm_source=test` and `utm_campaign=test` are **NOT** passed through to the form `fxa-email-form-utm-source` and `fxa-email-form-utm-campaign` hidden fields.
- [x] FxA flow params (`flow_id`, `device_id`) are **NOT** added to hidden fields.
- [x] When filling out the form, email **IS** still passed through to FxA as expected.
 
With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] `utm_source=test` and `utm_campaign=test` **ARE** passed through to the form `fxa-email-form-utm-source` and `fxa-email-form-utm-campaign` hidden fields.
- [x] FxA flow params (`flow_id`, `device_id`) **ARE** added to hidden fields.
- [x] When filling out the form, email **IS* still passed through to FxA as expected.

When using Firefox on desktop:

- [x] `<input type="hidden" name="context" value="fx_desktop_v3">` **IS** added to the form.

---

### VPN attribution (EU)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?geo=pl&utm_source=test&utm_campaign=test

#### With GPC/DNT enabled:

- [x] Consent banner is **NOT** shown. 
- [x] `utm_source=test&utm_campaign=test` are **NOT** propagated through to subscription links.
- [x] FxA flow parameters (`flow_id`, `device_id` etc) are **NOT** added to subscription links.

#### With GCP/DNT disabled:

- [x] Consent banner **IS** shown.
- [x] Clicking "Accept" closes banner. 
  - [x] `utm_source=test&utm_campaign=test` **ARE** propagated through to subscription links.
  - [x] FxA flow parameters (`flow_id`, `device_id` etc) **ARE** added to subscription links.
  - [x] Refreshing page persists consent choice and analytics behavior.
- [x] Clicking "Reject" closes banner.
  - [x] `utm_source=test&utm_campaign=test` are **NOT** propagated through to subscription links.
  - [x] FxA flow parameters (`flow_id`, `device_id` etc) are **NOT** added to subscription links.
  - [x] Refreshing page persists consent choice and analytics behavior.

---

### VPN attribution (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test

#### By default: (does not require DNT/GPC)

- [x] `utm_source=test&utm_campaign=test` **ARE** propagated through to subscription links.
- [x] FxA flow parameters (`flow_id`, `device_id` etc) **ARE** added to subscription links.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] `utm_source=test&utm_campaign=test` are **NOT** propagated through to subscription links.
- [x] FxA flow parameters (`flow_id`, `device_id` etc) are **NOT** added to subscription links.
 
With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] `utm_source=test&utm_campaign=test` **ARE** propagated through to subscription links.
- [x] FxA flow parameters (`flow_id`, `device_id` etc) **ARE** added to subscription links.

---

### VPN affiliate marketing flow (EU)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?geo=de&cjevent=123456789

#### With GPC/DNT enabled:

- [x] Consent banner is **NOT** shown.
- [x] Fetch request is **NOT** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
- [x] `moz-cj-affiliate` cookie is **NOT** set.

#### With GCP/DNT disabled:

- [x] Consent banner **IS** shown.
- [x] Clicking "Accept" closes banner
  - [x] Fetch request **IS** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
  - [x] `moz-cj-affiliate` cookie **IS** set.
  - [x] Refreshing page persists consent choice and analytics behavior.
- [x] Clicking "Reject" closes banner
  - [x] `moz-cj-affiliate` cookie is **NOT** set.
  - [x] Fetch request is **NOT** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
  - [x] Refreshing page persists consent choice and analytics behavior.
---

### VPN affiliate marketing flow (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?cjevent=123456789

#### With GPC/DNT enabled:

- [x] Fetch request is **NOT** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
- [x] `moz-cj-affiliate` cookie is **NOT** set.
 
#### With GCP/DNT disabled:

- [x] Fetch request **IS** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
- [x] `moz-cj-affiliate` cookie **IS** set.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Fetch request is **NOT** made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
- [x] `moz-cj-affiliate` cookie is **NOT** set. 

With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Fetch request **IS* made to `stage.cjms.nonprod.cloudops.mozgcp.net/aic`.
- [x] `moz-cj-affiliate` cookie **IS** set. 

---

### VPN / Stripe Radar (EU)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?geo=de

#### With GPC/DNT enabled:

- [x] Consent banner is **NOT** shown. 
- [x] Cookies from `js.stripe.com` are **NOT** set.

#### With GCP/DNT disabled:

- [x] Consent banner **IS** shown.
- [x] Clicking "Accept" closes banner. 
  - [x] Cookies from `js.stripe.com` **ARE** set.
  - [x] Refreshing page persists consent choice and analytics behavior.
- [x] Clicking "Reject" closes banner.
  - [x] Cookies from `js.stripe.com` are **NOT** set.
  - [x] Refreshing page persists consent choice and analytics behavior.

---

### VPN / Stripe Radar (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/products/vpn/

#### By default:

- [x] Cookies from `js.stripe.com` **ARE** set.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Cookies from `js.stripe.com` are **NOT** set.
 
With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Cookies from `js.stripe.com` **ARE** set.

---

### Glean (EU)

URL: https://www-demo1.allizom.org/en-US/products/vpn/?geo=de

#### With GPC/DNT enabled:

- [x] Consent banner is **NOT** shown.
- [x] Glean page load ping is **NOT** sent.

#### With GPC/DNT disabled:

- [x] Consent banner **IS** shown.
- [x] Clicking "Accept" closes banner
  - [x] Glean page load ping **IS** sent.
  - [x] Refreshing page persists consent choice and analytics behavior.
- [x] Clicking "Reject" closes banner
  - [x] Glean deletion request ping **IS** sent.
  - [x] Refreshing page should show no ping.

---

### Glean (rest-of-world)

URL: https://www-demo1.allizom.org/en-US/products/vpn/

#### By default (does not require DNT/GPC):

- [x] Glean page load ping **IS** sent.

With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Glean deletion request ping **IS** sent.

With a consent cookie set that [accepts analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):

- [x] Glean page load ping **IS** sent.

---

### Page banners (same logic applies for EU and rest-of-world)

This is not so easy to test as right now there is not a banner shown on desktop. The easiest way is to [change this code locally](https://github.com/mozilla/bedrock/blob/main/media/js/base/banners/firefox-app-store-banner-init.js) so the firefox app store banner gets shown on desktop.

URL: http://localhost:8000/en-US/

- [x] `firefox-app-store-banner` cookie **IS** set when clicking close (X) button on app banner.
- [x] Monitor banner is **NOT** displayed again on page reload.

With a consent cookie set that [rejects preference cookies](http://localhost:8000/en-US/privacy/websites/cookie-settings/):

- [x] `firefox-app-store-banner` cookie is **NOT** set when clicking close (X) button on app banner.
- [x] App banner **IS** displayed again on page reload.

With a consent cookie set that [accepts preference cookies](http://localhost:8000/en-US/en-US/privacy/websites/cookie-settings/):

- [x] `firefox-app-store-banner` cookie **IS** set when clicking close (X) button on app banner.
- [x] App banner is **NOT** displayed again on page reload.

---

### Traffic Cop (same logic applies for EU and rest-of-world)

URL: https://www-demo1.allizom.org/en-US/products/vpn/

#### With GPC/DNT enabled:

- [x] Experiment redirect is not triggered.

#### With GPC/DNT disabled:

- [x] Experiment redirects to https://www-demo1.allizom.org/en-US/products/vpn/?entrypoint_experiment=vpn-headlines&entrypoint_variation=1

---

### GPC (Global Privacy Control)

URL: https://www-demo1.allizom.org/.well-known/gpc.json

- [x] Returns valid `application/json` content type

alexgibson commented on media/js/base/consent/allow-list.es6.js:
```diff
@@ -0,0 +1,21 @@
+ */
+
+const MozAllowList = [
+    '/firefox/challenge-the-default/',
+    '/firefox/features/*',
+    '/firefox/more/*',
+    '/firefox/pocket/',
+    '/firefox/set-as-default/',
+    '/firefox/switch/',
+    '/firefox/unsupported-systems/',
+    '/products/',
+    '/products/mozsocial/invite/',
+    '/products/vpn/*',
+    '/try-picture-in-picture/'
```
TODO: this list is still TBD and subject to change.

alexgibson commented on bedrock/base/templates/includes/affiliate-banner.html:
Note for reviewer: this is being removed so we only have to maintain a single banner

alexgibson commented on media/js/base/banners/mozilla-banner.es6.js:
```diff
@@ -0,0 +1,141 @@
+};
+
+MozBanner.setCookie = function (id) {
+    /**
+     * When closing a banner we set a "preference cookie" to remember
+     * that a user no longer wishes to see it on repeat page visits.
+     * Legal are OK to set this cookie without explicit consent in
+     * EU/EAA countries because:
+     *
+     * 1) The cookie is not used for tracking purposes in any way.
+     * 2) The cookie does not leave the users’ device.
+     * 3) The cookie is set only after an explicit action by the user
+     *    that signals clear intent.
+     *
+     * We still honor not setting this cookie if preference cookies
+     * have been explicitly rejected by the user.
+     */
+    if (!MozBanner.consentsToCookies()) {
+        return;
+    }
```
Note for reviewer: most of this file is just existing code that has been re-formatted to modern JS. The only real new code is here.

alexgibson commented on media/js/base/banners/mozilla-banner.es6.js:
```diff
@@ -0,0 +1,140 @@
+};
+
+MozBanner.setCookie = function (id) {
+    /**
+     * When closing a banner we set a "preference cookie" to remember
+     * that a user no longer wishes to see it on repeat page visits.
+     * Legal are OK to set this cookie without explicit consent in
+     * EU/EAA countries because:
+     *
+     * 1) The cookie is not used for tracking purposes in any way.
+     * 2) The cookie is set only after an explicit action by the user
+     *    that signals clear intent.
+     *
+     * We still honor not setting this cookie if preference cookies
+     * have been explicitly rejected by the user.
+     */
+    if (!MozBanner.consentsToCookies()) {
+        return;
+    }
```
Note for reviewer: most of this file is just existing code that has been reformatted to modern JS. The only real new code added is here.

alexgibson commented on media/js/base/fxa-product-button.es6.js:
```diff
@@ -169,7 +169,7 @@ FxaProductButton.init = function (overwrite) {
     // Collect all Fxa product buttons
     _buttons = document.getElementsByClassName('js-fxa-product-button');
 
-    if (!FxaProductButton.isSupported() || _buttons.length === 0) {
```
Note for reviewer: removed this here since it is already covered by a check on line 177 below.

alexgibson commented on media/js/base/gtm/ie/gtm-snippet-ie-init.es6.js:
```diff
@@ -0,0 +1,70 @@
+/*
```
Note for reviewer: most of the code here is ported directly from https://github.com/mozilla/bedrock/blob/main/media/js/ie/gtm-snippet.js rather than being new.

craigcook commented on bedrock/privacy/templates/privacy/cookie-settings.html:
```diff
@@ -0,0 +1,196 @@
+<nav aria-label="breadcrumb" class="mzp-c-breadcrumb is-hidden">
+  <ol class="mzp-c-breadcrumb-list">
+    <li class="mzp-c-breadcrumb-item">
+      <a href="#">{{ ftl('cookie-settings-breadcrumb-link') }}</a>
```
Link missing?

alexgibson replied:
This is a link that returns people to their previous page, so it uses JS to go back a step in browser history.

stephaniehobson reviewed (commented):
Submitting what I've reviewed so far (about 10% done) so you can respond. 

I don't like doing this because sometimes I ask for changes to things I'd find were totally fine if I had looked at the entire PR. 🤷‍♀️

alexgibson commented on bedrock/base/templates/base-protocol.html:
```diff
@@ -125,18 +133,18 @@
     {% block stub_attribution %}
       {% if settings.STUB_ATTRIBUTION_RATE %}
-        <!--[if !IE]><!-->
+        <!--[if IE 9]><!-->
```
Yes.

- Firefox has not been supported on Windows XP and Windows 7 since 2016.
- We removed CSS support for IE8 in 2016.
- We removed JS support for IE8 in 2021 (barring this).

Given the above, in 2024 I do not think it makes sense for us to spend time and effort supporting analytics features for unsupported operating systems that are now pretty ancient and have a very small user base. The fact that we're still supporting IE9 here already goes further than I think we necessarily have to.

alexgibson commented on bedrock/mozorg/templates/mozorg/robots.txt:
```diff
@@ -6,5 +6,6 @@ disallow: /*/firstrun/
 disallow: /*/newsletter/existing/
 disallow: /*/whatsnew/
 disallow: /*/etc/
+disallow: /*/privacy/websites/cookie-settings/
```
This was discussed a bit early on, but we never really reached a solid opinion on it. I'm happy to remove as I agree searching for this page sounds useful.

alexgibson commented:
Thanks @stephaniehobson - pushed a commit with some review fixes

stephaniehobson reviewed (commented):
*Question:* Is it a concern that for **FxA form (rest-of-world)** if DNT/GPC are enabled the form behaves as if the user has _consented_ to analytics?

stephaniehobson reviewed (changes requested):
**Issue:** Banner remains visible after consent cookie is set.

Steps to reproduce:
1) Land on a page which opens the banner
2) Click `Cookie settings`
3) Configure cookies and save
4) Use the back button or click previous page

Expected behaviour:
- Consent has been configured, the banner does not display

Actual behaviour:
- Consent banner is still visible.
(reloading the page acts as expected and does make it respect the new settings)

I'm open to this being non-blocking if it can be addressed in a follow-up PR, however, this is how I commonly interact with these banners so there's at least one user out there would would be confused by it.

stephaniehobson reviewed (commented):
> With a consent cookie set that [rejects analytics](https://www-demo1.allizom.org/en-US/privacy/websites/cookie-settings/):
> 
> [ ] Glean deletion request ping IS sent.

I'm not sure what I'm looking for to confirm this.

alexgibson commented on media/js/base/consent/utils.es6.js:
```diff
@@ -0,0 +1,219 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+import MozAllowList from './allow-list.es6';
```
Allow-list has been the internal name that people have used to refer to this, so I'd rather keep it aligned to how we talk about it in code, but thanks for the suggestion.

alexgibson commented on media/js/base/consent/init.es6.js:
```diff
@@ -0,0 +1,170 @@
+            break;
+    }
+
+    // Global helper function to get consent state for debugging purposes.
```
Yes, my intent here is to keep this around for it / when we need to help people debug anything.

alexgibson commented:
> Question: Is it a concern that for FxA form (rest-of-world) if DNT/GPC are enabled the form behaves as if the user has consented to analytics?

Not at this time, but this could change in the future. The mandate thus far has been to fix issues for EU/EAA, with minimal changes for rest-of-world. From legal's perspective this is OK for now, but how we handle consent longer term for rest-of-world will likely be a followup item to address.

> Re Glean: I'm not sure what I'm looking for to confirm this.

See: https://bedrock.readthedocs.io/en/latest/attribution/0001-analytics.html#debugging-pings. You should be able to see when a deletion request ping is sent.

> Issue: Banner remains visible after consent cookie is set.

What you're seeing here is [bfcache](https://developer.mozilla.org/en-US/docs/Glossary/bfcache), which is default browser behavior. Basically when clicking the back button, the previous page is cached by the browser rather than requiring it to be loaded over the network again. There are ways to disable this, but the obvious drawback is we lose the performance gains of bfcache. I'll take a look to see if there's anything we can do here, but my gut is to not really start messing with this stuff.

janbrasna commented on media/css/privacy/cookie-settings-form.scss:
```diff
@@ -0,0 +1,149 @@
+        display: none;
+    }
+
+    .mzp-c-breadcrumb-list {
```
I'll look into fixing this properly upstream with the right tokens etc., all the instances deserve this.

Would you mind adding something like: _("TODO")_

```suggestion
    // Temporary fix until https://github.com/mozilla/protocol/issues/933
    .mzp-c-breadcrumb-list {
```

until this comes in base protocol to remind that it should be cleaned up?

(Or maybe higher on L10 at the block level.)

alexgibson commented:
@stephaniehobson I pushed a commit containing a potential fix for the bfcache issue you spotted. It makes me a little nervous to write code that will automatically refresh a web page, so let's only use this if you feel like the fix is worthwhile. I did add some guardrails around it so that the code will only run if a consent banner was visible on the previous page. Let me know what you think? I pushed it to the demo instance for testing.

```javascript
/**
 * If user clicks back button from /cookie-settings/ and
 * the page is in bfcache, reload the page if they have
 * set a consent cookie to avoid showing the banner again
 * inadvertently.
 * @param {Object} e - The page navigation event.
 */
function handleBfCacheNavigation(e) {
    if (e.persisted && hasConsentCookie()) {
        window.location.reload();
    }
}

window.addEventListener('pageshow', handleBfCacheNavigation, false);
```

stephaniehobson commented:
Glean (EU) > With GPC/DNT disabled > Clicking "Reject" closes banner >  Glean deletion request ping IS sent.

When I rejected cookies I saw this `(Glean.core.Pings.Maker) Storage for deletion-request empty. Ping will still be sent.`

stephaniehobson commented:
**Chore:** The following strings can be removed from `vpn/shared.ftl`:

`vpn-shared-affiliate-notification-message`
`vpn-shared-affiliate-notification-reject`
`vpn-shared-affiliate-notification-ok`

stephaniehobson reviewed (commented):
I have tested/reviewed all the functionality and files I wanted to. I think the only merge-blocking thing is that I have not been able to verify the Glean deletion request ping. That could still be tester error though.

Removing the banner when the user goes back to the banner page after viewing cookie settings is not a launch blocker IMHO (now that I know what it would take to fix it...) but we should make sure we're both comfortable with the current state before it merges. Maybe we can clear that up on Slack tomorrow or Monday morning.

alexgibson commented:
> When I rejected cookies I saw this `(Glean.core.Pings.Maker) Storage for deletion-request empty. Ping will still be sent.`

This is the expected result. The log message is worded a little oddly for sure. I asked the Glean folks about it, and they say it basically means the ping has no custom metrics attached to it other than the `client_info` section that contains `client_id` etc. This is all that's needed for a deletion-request since it basically tells our telemetry systems to "delete all data containing this client_id".

alexgibson commented:
Thanks for the detailed review @stephaniehobson! 

I think I've addressed all comments so far. I think we're both comfortable enough with `handleBfCacheNavigation()`? It seems to work well enough in testing, and should only fire under the specific condition when a banner is shown.

stephaniehobson commented on bedrock/mozorg/templates/mozorg/gpc.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "gpc": true,
+  "lastUpdate": "2024-03-26"
```
@alexgibson update this and I'll R+ you.

stephaniehobson commented:
R+ as soon as the [lastupdated date is updated](https://github.com/mozilla/bedrock/pull/14559#pullrequestreview-2048258163).  

Very through and thoughtful work on a massive change for the better. ✨

alexgibson commented:
Thanks @stephaniehobson - can you please give me an r+wc and then I can update the date and merge on Monday? Thank you

stephaniehobson reviewed (approved):
R+wc

Just needs the date updated.
